// @vuepress/plugin-palette/lib/node/palettePlugin.js
import { addViteConfig, chainWebpack } from '@vuepress/helper';
import { watch } from 'chokidar';
import { preparePaletteFile } from './preparePaletteFile.js';
import { prepareStyleFile } from './prepareStyleFile.js';
import { presetOptions } from './presetOptions.js';
/**
 * Create a palette plugin
 *
 * 创建调色板插件
 *
 * @param [options={}] - plugin options / 插件选项
 *
 * @example
 * ```ts
 * import { palettePlugin } from '@vuepress/plugin-palette'
 *
 * export default {
 *   plugins: [
 *     palettePlugin({
 *       preset: 'sass'
 *     })
 *   ]
 * }
 * ```
 */
export const palettePlugin = ({ preset = 'css', userPaletteFile = presetOptions[preset].userPaletteFile, tempPaletteFile = presetOptions[preset].tempPaletteFile, userStyleFile = presetOptions[preset].userStyleFile, tempStyleFile = presetOptions[preset].tempStyleFile, importCode = presetOptions[preset].importCode, } = {}) => ({
    name: '@vuepress/plugin-palette',
    alias: (app) => ({
        '@vuepress/plugin-palette/palette': app.dir.temp(tempPaletteFile),
        '@vuepress/plugin-palette/style': app.dir.temp(tempStyleFile),
    }),
    extendsBundlerOptions: (bundlerOptions, app) => {
        if (preset !== 'sass')
            return;
        // silent import deprecation for vite
        addViteConfig(bundlerOptions, app, {
            css: {
                preprocessorOptions: {
                    sass: {
                        silenceDeprecations: ['import'],
                    },
                    scss: {
                        silenceDeprecations: ['import'],
                    },
                },
            },
        });
        // silent import deprecation for webpack
        chainWebpack(bundlerOptions, app, (webpackOptions) => {
            webpackOptions.module
                .rule('scss')
                .use('sass-loader')
                .tap((loaderOptions) => ({
                ...loaderOptions,
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                sassOptions: {
                    ...loaderOptions.sassOptions,
                    silenceDeprecations: [
                        'import',
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
                        ...(loaderOptions.sassOptions?.silenceDeprecations ?? []),
                    ],
                },
            }));
        });
    },
    onPrepared: async (app) => {
        await Promise.all([
            preparePaletteFile(app, {
                userPaletteFile,
                tempPaletteFile,
                importCode,
            }),
            prepareStyleFile(app, {
                userStyleFile,
                tempStyleFile,
                importCode,
            }),
        ]);
    },
    onWatched: (app, watchers) => {
        const paletteWatcher = watch(userPaletteFile, {
            cwd: app.dir.source(),
            ignoreInitial: true,
        });
        paletteWatcher.on('add', () => {
            void preparePaletteFile(app, {
                userPaletteFile,
                tempPaletteFile,
                importCode,
            });
        });
        paletteWatcher.on('unlink', () => {
            void preparePaletteFile(app, {
                userPaletteFile,
                tempPaletteFile,
                importCode,
            });
        });
        watchers.push(paletteWatcher);
        const styleWatcher = watch(userStyleFile, {
            cwd: app.dir.source(),
            ignoreInitial: true,
        });
        styleWatcher.on('add', () => {
            void prepareStyleFile(app, {
                userStyleFile,
                tempStyleFile,
                importCode,
            });
        });
        styleWatcher.on('unlink', () => {
            void prepareStyleFile(app, {
                userStyleFile,
                tempStyleFile,
                importCode,
            });
        });
        watchers.push(styleWatcher);
    },
});
